private/tools/java/com/github/bazelbuild/rules_jvm_external/maven/MavenPublisher.java

```diff
@@ -305,6 +378,83 @@ private static String toHexS(String fmt, String algorithm, byte[] toHash) {
     }
   }
 
+  /**
+   * Attempts to download the file at the given targetUrl. Valid protocols are: http(s), file, and
+   * s3 at the moment.
+   */
+  private static CompletableFuture<Optional<String>> download(String targetUrl) {
+    if (targetUrl.startsWith("http")) {
```

Comment:
Why not do a `URI.create(targetUrl)` and do a switch on the scheme? You don't want to use `URL` because that does a DNS lookup in its constructor, but `URI` should be safe.